zmq: mempool notifications

laanwj opened the pull request:
Add notifications when transactions enter or leave the mempool.

These can be enabled with `zmqpubmempool`:
- `mempooladded`: a transaction was added to the mempool
- `mempoolremoved`: a transaction was removed from the mempool

This allows third-party software to keep track of what is in the mempool in real time. For example a mempool monitor I'm working on:
![schermafdruk van 2016-03-27 17-42-16](https://cloud.githubusercontent.com/assets/126646/14066461/eccfffbe-f449-11e5-843e-1295b7e85f46.png)
It also makes it possible to keep statistics on why transactions disappear from the mempool.

Full documentation: https://github.com/bitcoin/bitcoin/blob/325afe68381d59f1b95a690927c4a8ea886ac791/doc/zmq.md#mempooladd

See individual commits for details.

Current issues:
- Warning `NotifyEntryAdded: WARNING: zmq send failed with code -1` spam in log if "-zmqpubmempool" is not enabled.

laanwj committed:
mempool: add notification for added/removed entries

Add notification signals to make it possible to subscribe to mempool
changes:

- NotifyEntryAdded(const CTxMemPoolEntry &)>
- NotifyEntryRemoved(const CTxMemPoolEntry &, MemPoolRemovalReason)>

Also add a mempool removal reason enumeration, which is passed to the
removed notification based on why the transaction was removed from
the mempool.

laanwj committed:
zmq: refactor initialization

Replace factories list with calls to a function.

This simplifies the code (every notifier is only created once anyway),
and makes it possible to pass arguments to notifier contructors.

laanwj committed:
zmq: add mempool notifications

Add notifications when transactions enter or leave the mempool.

These can be enabled with `pubmempool`:

- `mempooladded`: a transaction was added to the mempool
- `mempoolremoved`: a transaction was removed from the mempool

This allows third-party software to keep track of what is in the mempool
in real time.

laanwj committed:
doc: Add documentation for zmq mempool notifications

promag commented:
I was planning to submit a pull request to change this kind of *notifications*. The current solution is not very elegant and doesn't scale.

So if you don't mind I'll try to explain here the idea using this example:

Usage: `-notify "mempool ! zmqpub address=tcp://127.0.0.1:28332"`

Here the notify argument creates one notification *pipeline* composed by one *source* of events (mempool) and one *sink* of events (zmqpub), connected with the operator `!`.

dcousens commented:
concept ACK

laanwj commented:
@promag I'm all for making it more flexible, but what you're proposing sounds like the gstreamer syntax and that's IMO even harder to use than the current arguments. I've needed to do a few media pipelines in my PhD research but never managed to really master it.

Our basic idea with the ZMQ notifications is that **a** notification system exists. From there on everything can be handled by external software. I don't want too much notification-related complexity in core.

Aside, this is the wrong place to discuss this. Let's focus on the software change here. Please create a new issue.

promag commented:
@laanwj yes gstreamer all over the comment. I'll move the discussion to a new issue.

utACK.

promag commented on src/init.cpp:
```diff
@@ -1230,7 +1230,7 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
 
 #if ENABLE_ZMQ
-    pzmqNotificationInterface = CZMQNotificationInterface::CreateWithArguments(mapArgs);
+    pzmqNotificationInterface = CZMQNotificationInterface::CreateWithArguments(mapArgs, &mempool);
```
Feels a bit hacky.

laanwj replied:
Better suggestions?
As I see ti there are two alternatives:
- use the global mempool object (ugly)
- pass the mempool in (slightly les ugly)

jonasschnelli replied:
I think this implementation approach is reasonable and I also don't see a more elegant way to do this.

promag replied:
Then I would rename `CreateWithArguments` to just `Create` and move the following from src/init.cpp to `Create`.

```
if (pzmqNotificationInterface) {
    RegisterValidationInterface(pzmqNotificationInterface);
}
```

laanwj replied:
I agree, will move that line.

jonasschnelli commented:
Awesome!
Nice work! I hope you have also plans to opensource the ncurses UI.

Some ideas / conceptual questions:
- How could a listener keep track of the overall mempool size without listening to a fresh start (listening from the beginning when the mempool was at a size of 0), also there is a risk of a delta in case you missed a notification in case you calculate the mempool overall size on the "listeners side".

- would a notification after a significant (every X MB, every X transaction added/removed) in the mempool size or structure make sense?

- Adding the total mempool size in every mempool notification is probably to much?

jonasschnelli commented:
Code Review utACK.

promag commented:
> Adding the total mempool size in every mempool notification is probably to much?

:+1:

laanwj commented:
> How could a listener keep track of the overall mempool size without listening to a fresh start (listening from the beginning when the mempool was at a size of 0), also there is a risk of a delta in case you missed a notification in case you calculate the mempool overall size on the "listeners side".

What if you use the following sequence:
- Start listening (queue up events)
- Request `getrawmempool`
- Start applying events

You may have some duplicates in the first run, e.g., adds which already show up in the mempool requested, removes which were already removed, but I think due to the nature of the events (just add and remove) the overall state will be the same?

>  there is a risk of a delta in case you missed a notification

Missing notifications are a possibiltiy with zmq. But I'd say this is a matter of sizing buffers appropriately, and dedicating a thread to listening to zmq requests.

To reliably detect missing notifications we could add sequence numbers to the events, as I think zmq guarantees that events from a single source are delivered in order.

Then again if we do this we probably want it for all the event types...

> Adding the total mempool size in every mempool notification is probably to much?

I thought about sending a generic stats packet with every mempool event, but as you can receive tons of them I don't want to make the individual events too big and include things you can compute yourself. But no problems with just the size / # transactions.

jonasschnelli commented:
Tested ACK 4e3039b41946ad7bbb1aff6b6df5b1e83348c942
I also think the `addNotifier()`-refactor makes it much cleaner.

paveljanik commented:
lightly tested ACK https://github.com/bitcoin/bitcoin/commit/4e3039b41946ad7bbb1aff6b6df5b1e83348c942

sipa commented on src/txmempool.h:
```diff
@@ -319,6 +321,18 @@ class CInPoint
+/** Reason why a transaction was removed from the mempool,
+ * this is passed to the notification signal.
+ */
+enum MemPoolRemovalReason {
```
This enum seems to be part of the ZMQ API, so we'll probably want some stability across versions. Things like MPR_EXPIRY and MPR_SIZELIMIT seem quite specific to the existing implementation.

sipa commented on src/zmq/zmqpublishnotifier.cpp:
```diff
@@ -171,3 +177,41 @@ bool CZMQPublishRawTransactionNotifier::NotifyTransaction(const CTransaction &tr
+    LogPrint("zmq", "zmq: mempool entry removed: %s, reason %i\n", entry.GetTx().GetHash().ToString(), (int)reason);
+    uint8_t data[32 + 1];
+    WriteHash(&data[0], entry.GetTx().GetHash());
+    data[32] = reason;
```
I would prefer it if the value of the (internal) MemPoolRemovalReason did not leak into the ZMQ protocol, though that can be fixed later by introducing a new enum in the ZMQ definition.

laanwj replied:
Yes, I'll add another enum here and add a mapping.

laanwj commented:
Ok added two commits:
- The first adds a (currently, one to one) mapping between the mempool reasons from the mempool and those on the ZMQ interface.
- The second is a slight refactoring to make all ZMQ publish notifiers register (and unregister) themselves for the events that they want to receive, instead of relying on the events being propagated manually though `CZMQNotificationInterface` @promag

laanwj committed:
zmq: refactor to make all notifiers register themselves

Make the registration process less circuitous by making
the notifiers listen to their appropriate events themselves

laanwj commented:
Rebased after #7762

MarcoFalke commented:
Travis failure seems unrelated

```
Your test run exceeded 50 minutes. 
One possible solution is to split up your test run.

laanwj commented:
Rebased, and changed the enums (both for the mempool and for the zmq protocol) to c++11 enums.

paveljanik commented:
lightly tested ACK https://github.com/bitcoin/bitcoin/pull/7753/commits/1bbe3660ad9a375e2d2bb8c7fa8f38132774b656

sipa commented:
Needs rebase.

jonasschnelli commented:
Still needs rebase.

jmcorgan commented:
ACK.  I have a rebased version of this branch at 'jmcorgan/2016_03_zmq_mempool_notifications'.

jonasschnelli commented:
@jmcorgan feel free to open a PR of your rebases version and take over the lead.

jmcorgan commented:
Continued in #8549

laanwj commented:
Closing in favor of #8549
